src/primitives.js

```diff
@@ -212,50 +212,334 @@ export default {
 
 		return cuboids;
 	},
-	cylinder: ({ color, coords, opacity = 1, radius }, extent) => {
-		const cylinders = new InstancedMesh(
+	cylinder: ({ color, coords, edgeForm = {}, opacity = 1, radius = 1 }, extent) => {
+		// number of vertex per coordinate / number of coordinates per cylinder = 3 / 2
+		const cylindersBegin = new Float32Array(coords.length * 1.5);
+		const cylindersEnd = new Float32Array(coords.length * 1.5);
+
+		for (let i = 0; i < coords.length / 2; i++) {
+			const startCoordinate = coords[i * 2][0] ?? scaleCoordinate(coords[i * 2][1], extent);
+
+			const endCoordinate = coords[i * 2 + 1][0] ?? scaleCoordinate(coords[i * 2 + 1][1], extent);
+
+			cylindersBegin[i * 3] = startCoordinate[0];
```

Comment:
A suggestion is to make these 3 lines into a function which can be used twice here and also in the polyhedron code. And probably it would make it easier to understand if the function has a useful name like appendCoordinate or something like that.

Reply:
I don't think that copy something to the buffer should be a function, the thing you want to copy can be a [Vector3](https://threejs.org/docs/index.html?q=Vector#api/en/math/Vector3), an array, or just three values. What do you think of add a comment to that?

Reply:
You mean add a more generalized copy function? That would be good too. 

The higher level idea is that you seem to be accumulating coordinate values from a  3D point structure into a linear array and this kind.  And the fact that is done several times indicates something about this linear structure is is what in OO would be some sort of Object that has common operations associated with it.  

It is not clear what those common operations are other than somehow this is going to be the edge lines that will eventually get produced.

Reply:
> You mean add a more generalized copy function? That would be good too.

Sorry, I wasn't clear. I think a generalized copy function is the unique way to have a copy function, but a generalized function is [I think] worse than 3 lines of copying.

> It is not clear what those common operations are other than somehow this is going to be the edge lines that will eventually get produced.

This isn't related to the position attribute of the edges. The "position" attribute of the cylinders and edges is the vertices' position of a cylinder starting in (0, 0, 0). The serialized coordinates are transformed to get the position of the vertices of each cylinder and edges.

```js
// position, rotate and scale the cylinder
mat4 cylinderMatrix = mat4(x, 0, y, 0, z * height, 0, cylinderBegin, 1);
```
```js
// position, rotate and scale the edges
mat4 cylinderMatrix = mat4(x, 0, y, 0, z * height, 0, cylinderBegin, 1);
```

> The higher level idea is that you seem to be accumulating coordinate values from a 3D point structure into a linear array and this kind. And the fact that is done several times indicates something about this linear structure (a particular representation of a collection of points) is is what in OO would be some sort of Object that has common operations associated with it.

The coordinates need to be serialized to be copied to the GPU RAM (otherwise that would be crazily slow).

Reply:
> > You mean add a more generalized copy function? That would be good too.
> 
> Sorry, I wasn't clear. I think a generalized copy function is the unique way to have a copy function, but a generalized function is [I think] worse than 3 lines of copying.
> 

Right now it is 9 lines of copying. 3 groups of 3. 

> > It is not clear what those common operations are other than somehow this is going to be the edge lines that will eventually get produced.
> 
> This isn't related to the position attribute of the edges. The "position" attribute of the cylinders and edges is the vertices' position of a cylinder starting in (0, 0, 0). The serialized coordinates are transformed to get the position of the vertices of each cylinder and edges.
> 
> ```js
> // position, rotate and scale the cylinder
> mat4 cylinderMatrix = mat4(x, 0, y, 0, z * height, 0, cylinderBegin, 1);
> ```
> 
> ```js
> // position, rotate and scale the edges
> mat4 cylinderMatrix = mat4(x, 0, y, 0, z * height, 0, cylinderBegin, 1);
> ```
> 
> > The higher level idea is that you seem to be accumulating coordinate values from a 3D point structure into a linear array and this kind. And the fact that is done several times indicates something about this linear structure (a particular representation of a collection of points) is is what in OO would be some sort of Object that has common operations associated with it.
> 
> The coordinates need to be serialized to be copied to the GPU RAM (otherwise that would be crazily slow, and when it can be done slow a lot of people do it slow, that's why WebGL/OpenGL/probably all graphics APIs don't allow that).

If that is the motivation for this messiness, then state it somewhere so that 10 years down the line when we realize as in sleeper that tobacco, greasy food are good for you and GPUs are bad for you, someone will understand that this code should be adjusted.